runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/util/Convenience.kt

```diff
@@ -36,3 +37,36 @@ public fun truthiness(value: Any?): Boolean = when (value) {
     null -> false
     else -> true
 }
+
+public data class Property(val name: String, val value: Any?)
+
+@InternalApi
+@Suppress("UNCHECKED_CAST")
+public fun Any.getProperties(): List<Property> =
+    this::class.declaredMemberProperties.map {
+        Property(name = it.name, value = (it as KProperty1<Any, *>).get(this))
+    }
+
+@InternalApi
+public fun HashMap<String, Any?>.getProperties(): List<Property> =
+    this.map {
+        Property(it.key, it.value)
+    }
```

Comment:
**Correctness**: In general, we should avoid reflection at runtime. It's not necessarily performant, may pick up additional members beyond the ones we intend, and requires the use of an additional runtime dependency (**kotlin-reflect**) which bloats deployment sizes.

I think we should explore other means of getting these data at runtime besides reflection. Since the objects we need keys/values for are defined in the model, we can use codegen to get more intelligent data about the them. I think there are a few options, one of which is something like:
```kotlin
// Code generated by smithy-kotlin-codegen. DO NOT EDIT!
class Bucket {
    val name = ...
    val creationDate = ...

    companion object {
        // Codegen each field as an `AttributeKey`:
        val Name = AttributeKey<String>("name")
        val CreationDate = AttributeKey<Instant>("creationDate")

        // Also codegen an `AllFields` list:
        val AllFields = listOf(
            Name,
            CreationDate,
        )
    }

    // Then codegen a method to get the attributes themselves:
    fun toAttributes() = attributesOf {
        Name to name,
        CreationDate to creationDate,
    }
}
```

There are a few nice properties to this:
* Avoids reflection, obvs
* Makes it so that values are retrievable as specific types rather than `Any?`
  * For example, `bucket.toAttributes()[Bucket.Name]` is a `String?`
* This _may_ enable more advanced use cases down the road related to non-SDK serialization

The downsides here are:
* Increases artifact size since we're adding new fields/methods to _every_ structure/union shape
* Codegen is tricky, there could be unforeseen naming collisions
  * Using a method named `toAttributes` rather than a property called `attributes` should help with that but still

There are probably other ways to solve this problem without reflection too.

Reply:
It sounds like we should talk about this more. I 100% agree with Ian that we probably want to avoid reflection. If we are generating code like this I have concerns and think this warrants a larger discussion on how we solve this problem. Can we maybe put together a quip of the problem and our options to explore? I wouldn't want to generate this for every type since the only use case right now is for unused waiter functions and this will result in a different kind of bloat from codegen of unused things.